agent/backend/otel/utils.go

```diff
@@ -0,0 +1,96 @@
+package otel
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/netboxlabs/orb-agent/agent/backend"
+)
+
+func (d *openTelemetryBackend) getProcRunningStatus() (backend.RunningStatus, string, error) {
+	if d.proc == nil {
+		return backend.Unknown, "backend not started yet", nil
+	}
+	status := d.proc.Status()
+
+	if status.Error != nil {
+		errMsg := fmt.Sprintf("device-discovery process error: %v", status.Error)
+		return backend.BackendError, errMsg, status.Error
+	}
+
+	if status.Complete {
+		err := d.proc.Stop()
+		return backend.Offline, "device-discovery process ended", err
+	}
+
+	if status.StopTs > 0 {
+		return backend.Offline, "device-discovery process ended", nil
+	}
+	return backend.Running, "", nil
+}
+
+// note this needs to be stateless because it is called for multiple go routines
+func (d *openTelemetryBackend) request(url string, payload interface{}, method string, body io.Reader, contentType string, timeout int32) error {
+	client := http.Client{
+		Timeout: time.Second * time.Duration(timeout),
+	}
+
+	status, _, err := d.getProcRunningStatus()
+	if status != backend.Running {
+		d.logger.Warn("skipping device discovery REST API request because process is not running or is unresponsive", zap.String("url", url), zap.String("method", method), zap.Error(err))
+		return err
+	}
+
+	URL := fmt.Sprintf("%s://%s:%s/api/v1/%s", d.apiProtocol, d.apiHost, d.apiPort, url)
+
+	req, err := http.NewRequest(method, URL, body)
+	if err != nil {
+		d.logger.Error("received error from payload", zap.Error(err))
+		return err
+	}
+
+	req.Header.Add("Content-Type", contentType)
+	res, getErr := client.Do(req)
+
+	if getErr != nil {
+		d.logger.Error("received error from payload", zap.Error(getErr))
+		return getErr
+	}
+
+	defer func() {
+		if err := res.Body.Close(); err != nil {
+			d.logger.Error("failed to close response body", zap.Error(err))
+		}
+	}()
+
+	if (res.StatusCode < 200) || (res.StatusCode > 299) {
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			return fmt.Errorf("non 2xx HTTP error code from device-discovery, no or invalid body: %d", res.StatusCode)
+		}
+		if len(body) == 0 {
+			return fmt.Errorf("%d empty body", res.StatusCode)
+		} else if body[0] == '{' {
+			var jsonBody map[string]interface{}
+			err := json.Unmarshal(body, &jsonBody)
+			if err == nil {
+				if errMsg, ok := jsonBody["message"]; ok {
+					return fmt.Errorf("%d %s", res.StatusCode, errMsg)
+				}
+			}
```

Comment:
It seems like there are cases where this can drop through without returning an error. Not sure if that's intentional. 

eg if json unmarhaling fails here or there was a body that didn't start with '{', should it still return an error ?